pkg/distributor/distributor_test.go

```diff
@@ -1467,6 +1467,244 @@ func TestDistributor_Push_HistogramValidation(t *testing.T) {
+						value++
+					}
+					samples[i].TimestampMs = timestamp
+					samples[i].Value = float64(value)
+				}
+				timeseries := []mimirpb.PreallocTimeseries{
+					makeTimeseries(labels, samples, nil, nil),
+				}
+				return timeseries
+			},
+
+			expectedSamples: []mimirpb.PreallocTimeseries{
+				makeTimeseries(labels, makeSamples(timestamp, 1.0), nil, nil),
+			},
+			expectedErrors: []error{
+				fmt.Errorf("samples with duplicated timestamps have been discarded, discarded samples: %d series: '%.200s' (err-mimir-sample-duplicate-timestamp)", 79_999, metricName),
+			},
+		},
+		"one timeseries with 80_000 histograms with duplicated timestamps": {
+			setup: func() []mimirpb.PreallocTimeseries {
+				histograms := make([]mimirpb.Histogram, testSize)
+				value := 0
+				for i := 0; i < testSize; i++ {
+					if i < numberOfDifferentValues {
+						value++
+					}
+					histograms[i] = makeHistograms(timestamp, generateTestHistogram(value))[0]
+				}
+				timeseries := []mimirpb.PreallocTimeseries{
+					makeTimeseries(labels, nil, histograms, nil),
+				}
+				return timeseries
+			},
+
+			expectedSamples: []mimirpb.PreallocTimeseries{
+				makeTimeseries(labels, nil, makeHistograms(timestamp, generateTestHistogram(1)), nil),
+			},
+			expectedErrors: []error{
+				fmt.Errorf("samples with duplicated timestamps have been discarded, discarded samples: %d series: '%.200s' (err-mimir-sample-duplicate-timestamp)", 79_999, metricName),
+			},
+		},
+		"one timeseries with 80_000 samples and 80_000 histograms with duplicated timestamps": {
+			setup: func() []mimirpb.PreallocTimeseries {
+				samples := make([]mimirpb.Sample, testSize)
+				histograms := make([]mimirpb.Histogram, testSize)
+				value := 0
+				for i := 0; i < testSize; i++ {
+					if i < numberOfDifferentValues {
+						value++
+					}
+					samples[i].TimestampMs = timestamp
+					samples[i].Value = float64(value)
+					histograms[i] = makeHistograms(timestamp, generateTestHistogram(value))[0]
+				}
+
+				timeseries := []mimirpb.PreallocTimeseries{
+					makeTimeseries(labels, samples, histograms, nil),
+				}
+				return timeseries
+			},
+
+			expectedSamples: []mimirpb.PreallocTimeseries{
+				makeTimeseries(labels, makeSamples(timestamp, 1.0), makeHistograms(timestamp, generateTestHistogram(1)), nil),
+			},
+			expectedErrors: []error{
+				fmt.Errorf("samples with duplicated timestamps have been discarded, discarded samples: %d series: '%.200s' (err-mimir-sample-duplicate-timestamp)", 159_998, metricName),
+			},
+		},
+	}
+
+	limits := prepareDefaultLimits()
+	ds, _, regs, _ := prepare(b, prepConfig{
+		limits:          limits,
+		numDistributors: 1,
+	})
+
+	// Pre-condition check.
+	require.Len(b, ds, 1)
+	require.Len(b, regs, 1)
+
+	for name, tc := range testCases {
+		b.Run(name, func(b *testing.B) {
+			b.ReportAllocs()
+			b.ResetTimer()
+			for n := 0; n < b.N; n++ {
+				b.StopTimer()
+				timeseries := tc.setup()
+				b.StartTimer()
```

Comment:
I'm not an expert on micro benchmarking in golang, but I'm not sure that this is a zero-sum game. The problem is that there's a stop-of-the-world for each `StartTimer()` and `StopTimer()`, and calling it in a tight loop may actually get measurements a bit skewed. What I typically do instead is to prepare `b.N` requests inside `b.Run()` then I call `b.ResetTimer()` and I run all `b.N` tests, using all the previously prepared requests.

Reply:
Thank you, this was a great tip.